gqlopentelemetry/tracer.go

```diff
@@ -0,0 +1,131 @@
+package gqlopentelemetry
+
+import (
+	"context"
+	"fmt"
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/label"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/99designs/gqlgen/graphql/handler/extension"
+)
+
+type Tracer struct {
+	ComplexityExtensionName string
+	DataDog                 bool
+}
+
+var _ interface {
+	graphql.HandlerExtension
+	graphql.ResponseInterceptor
+	graphql.FieldInterceptor
+} = Tracer{}
+
+func (a Tracer) ExtensionName() string {
+	return "OpenTelemetry"
+}
+
+func (a Tracer) Validate(schema graphql.ExecutableSchema) error {
+	return nil
+}
+
+func (a Tracer) InterceptResponse(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
+	tracer := global.Tracer("github.com/99designs/gqlgen-contrib")
+	ctx, span := tracer.Start(ctx, operationName(ctx))
+	defer span.End()
+	if !span.IsRecording() {
+		return next(ctx)
+	}
+
+	oc := graphql.GetOperationContext(ctx)
+	span.SetAttributes(
+		label.String("request.query", oc.RawQuery),
+	)
+	complexityExtension := a.ComplexityExtensionName
+	if complexityExtension == "" {
+		complexityExtension = "ComplexityLimit"
+	}
+	complexityStats, ok := oc.Stats.GetExtension(complexityExtension).(*extension.ComplexityStats)
+	if !ok {
+		// complexity extension is not used
+		complexityStats = &extension.ComplexityStats{}
+	}
+
+	if complexityStats.ComplexityLimit > 0 {
+		span.SetAttributes(
+			label.Int64("request.complexityLimit", int64(complexityStats.ComplexityLimit)),
+			label.Int64("request.operationComplexity", int64(complexityStats.Complexity)),
+		)
+	}
+
+	for key, val := range oc.Variables {
+		span.SetAttributes(
+			label.String(fmt.Sprintf("request.variables.%s", key), fmt.Sprintf("%+v", val)),
+		)
+	}
+
+	return next(ctx)
+}
+
+func (a Tracer) InterceptField(ctx context.Context, next graphql.Resolver) (interface{}, error) {
+	fc := graphql.GetFieldContext(ctx)
+	tracer := global.Tracer("")
```

Comment:
@k-yomo Do you know if we need to set name here?

Reply:
There is no other places to use the tracer name, so I think we don't need to make it const.